File: overlay/widgets/sector_timing.py
"""
Sector / lap timing -- current, last and best lap with live sector splits.

iRacing doesn't stream finished sector times directly, so SectorTimer derives
them from the player's lap-distance crossings: each time you pass a sector
boundary it records the split, tracks your best per sector, and rolls the lap
when you cross the start/finish line. The widget shows the running lap time, the
last/best laps, and a row of sector cells colored purple when you've just matched
your best for that sector.

Widget data dict (built by SectorTimer.snapshot()):
    cur_lap, last_lap, best_lap, predicted_lap
    sectors   list of {"time", "status", "active", "delta"} per sector
"""

from __future__ import annotations

from PyQt6.QtCore import QRectF, Qt
from PyQt6.QtGui import QColor, QPainter, QPen
from PyQt6.QtWidgets import QSizePolicy, QWidget

from .. import config
from .chrome import (cell_radius, col, draw_card, draw_dark_cell, draw_edge_band,
                     draw_metric_row, panel_pad)
from .fonts import data_font_bold, tabfont, tfont
from .formats import clock, sec, signed_delta

_SECTION = "sector_timing"


class SectorTimer:
    """Derives sector splits from lap-distance crossings (owned by the app)."""

    def __init__(self):
        self.starts: list[float] | None = None
        self.cur: list = []          # completed splits for the in-progress lap
        self.last: list = []         # previous lap's splits
        self.best: list = []         # best split seen per sector (personal)
        self.session_best: list = []  # fastest sector this session
        self.idx = 0                 # current sector index
        self._seg_start_t = 0.0      # lap time when the current sector began
        self._cur_lap_t = 0.0
        self._prev_pct = None

    def set_boundaries(self, starts) -> None:
        """Set sector start percentages (e.g. from SplitTimeInfo). Re-inits state
        if they changed (new track / session)."""
        if not starts:
            return
        s = sorted(float(x) for x in starts)
        if s != self.starts:
            self.starts = s
            self.cur, self.idx, self._seg_start_t, self._prev_pct = [], 0, 0.0, None

    def reset_session(self) -> None:
        self.session_best = []

    def update(self, pct, cur_lap_time, last_lap_time) -> None:
        if self.starts is None:
            self.starts = [i / 3.0 for i in range(3)]  # default 3 equal sectors
        if not isinstance(pct, (int, float)) or pct < 0:
            return
        if isinstance(cur_lap_time, (int, float)):
            self._cur_lap_t = cur_lap_time
        n = len(self.starts)

        # Lap rollover: big backward jump in lap distance = crossed start/finish.
        if self._prev_pct is not None and pct + 0.5 < self._prev_pct:
            if isinstance(last_lap_time, (int, float)) and last_lap_time > 0:
                self._finish_sector(last_lap_time - self._seg_start_t)
            self.last = self.cur[:] if self.cur else self.last
            self.cur, self.idx, self._seg_start_t = [], 0, 0.0
            self._prev_pct = pct
            return

        new_idx = 0
        for i in range(n):
            if pct >= self.starts[i]:
                new_idx = i
        if new_idx > self.idx and isinstance(cur_lap_time, (int, float)):
            self._finish_sector(cur_lap_time - self._seg_start_t)
            self._seg_start_t = cur_lap_time
            self.idx = new_idx
        self._prev_pct = pct

    def _finish_sector(self, t) -> None:
        i = len(self.cur)
        if not isinstance(t, (int, float)) or t <= 0:
            self.cur.append(None)
            return
        self.cur.append(t)
        while len(self.best) <= i:
            self.best.append(None)
        if self.best[i] is None or t < self.best[i]:
            self.best[i] = t
        while len(self.session_best) <= i:
            self.session_best.append(None)
        if self.session_best[i] is None or t < self.session_best[i]:
            self.session_best[i] = t

    def predicted_lap(self) -> float | None:
        """Sum of best sectors + current sector pace."""
        n = len(self.starts or [])
        if n <= 0:
            return None
        total = 0.0
        have = False
        for i in range(n):
            if i < len(self.cur):
                t = self.cur[i]
                if isinstance(t, (int, float)) and t > 0:
                    total += t
                    have = True
            elif i == self.idx:
                running = max(0.0, (self._cur_lap_t or 0.0) - self._seg_start_t)
                if running > 0:
                    total += running
                    have = True
            else:
                ref = None
                if i < len(self.session_best) and self.session_best[i]:
                    ref = self.session_best[i]
                elif i < len(self.best) and self.best[i]:
                    ref = self.best[i]
                if isinstance(ref, (int, float)) and ref > 0:
                    total += ref
                    have = True
        return total if have else None

    def snapshot(self, cur_lap, last_lap, best_lap, *, show_delta=False) -> dict:
        n = len(self.starts or [0, 0, 0])
        sectors = []
        for i in range(n):
            delta = None
            if i < len(self.cur):
                t = self.cur[i]
                best = self.best[i] if i < len(self.best) else None
                status = ("best" if (t is not None and best is not None
                                     and t <= best + 1e-6) else "done")
                if show_delta and t is not None and best is not None:
                    delta = t - best
                sectors.append({"time": t, "status": status, "active": False,
                                "delta": delta})
            elif i == self.idx:
                running = max(0.0, (self._cur_lap_t or 0.0) - self._seg_start_t)
                best = self.best[i] if i < len(self.best) else None
                if show_delta and best is not None and running > 0:
                    delta = running - best
                sectors.append({"time": running, "status": "running",
                                "active": True, "delta": delta})
            else:
                last = self.last[i] if i < len(self.last) else None
                sectors.append({"time": last, "status": "idle", "active": False,
                                "delta": None})
        return {
            "cur_lap": cur_lap,
            "last_lap": last_lap,
            "best_lap": best_lap,
            "predicted_lap": self.predicted_lap(),
            "sectors": sectors,
            "active_idx": self.idx,
        }


class SectorTimingWidget(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.data: dict = {}
        self.setMinimumSize(220, 120)
        self.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Expanding)

    def set_data(self, data: dict) -> None:
        data = data or {}
        if data == self.data:
            return
        self.data = data
        self.update()

    def paintEvent(self, event) -> None:  # noqa: N802
        config.use_section(_SECTION)
        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing)
        w, h = float(self.width()), float(self.height())
        d = self.data or {}
        cfg = config.CFG.get(_SECTION, {})
        card, radius = draw_card(p, w, h, _SECTION)
        data_bold = data_font_bold(_SECTION)

        pad = panel_pad(h)
        iw = card.width() - 2 * pad
        cur_h = h * 0.26 if cfg.get("show_predicted_lap") else h * 0.30
        p.setFont(tabfont(cur_h, bold=data_bold))
        p.setPen(col("text", _SECTION))
        p.drawText(QRectF(card.left() + pad, card.top() + pad, iw, cur_h),
                   Qt.AlignmentFlag.AlignCenter, clock(d.get("cur_lap")))
        if cfg.get("show_predicted_lap") and d.get("predicted_lap"):
            p.setFont(tfont(h * 0.11, bold=False))
            p.setPen(col("muted", _SECTION))
            p.drawText(QRectF(card.left() + pad, card.top() + pad + cur_h * 0.85,
                              iw, h * 0.08),
                       Qt.AlignmentFlag.AlignCenter,
                       f"Pred {clock(d.get('predicted_lap'))}")

        sub_top = card.top() + pad + (h * 0.34 if cfg.get("show_predicted_lap") else h * 0.30)
        fixed_rh = float(cfg.get("row_height_px", 0) or 0)
        if fixed_rh > 0:
            sub_h = fixed_rh
        else:
            sub_h = h * 0.18
            max_frac = float(cfg.get("max_row_height_frac", 0) or 0)
            if max_frac > 0:
                sub_h = min(sub_h, h * max_frac)
        sub_h = max(18.0, sub_h)
        sub = QRectF(card.left() + pad, sub_top, iw, sub_h)
        draw_edge_band(p, sub, "header_bg", _SECTION, bottom_line=True)
        half = sub.width() / 2
        self._pair(p, QRectF(sub.left(), sub.top(), half, sub.height()),
                   "LAST", clock(d.get("last_lap")), data_bold)
        self._pair(p, QRectF(sub.left() + half, sub.top(), half, sub.height()),
                   "BEST", clock(d.get("best_lap")), data_bold)

        sectors = d.get("sectors") or []
        if sectors:
            top = sub.bottom() + h * 0.04
            ch = card.bottom() - pad - top
            gap = iw * 0.03
            cw = (iw - gap * (len(sectors) - 1)) / len(sectors)
            x = card.left() + pad
            show_delta = cfg.get("show_sector_delta", False)
            for i, s in enumerate(sectors):
                self._cell(p, QRectF(x, top, cw, ch), i + 1, s, data_bold,
                           show_delta=show_delta)
                x += cw + gap

    def _pair(self, p, rect, label, value, data_bold) -> None:
        draw_metric_row(p, rect.adjusted(10, 0, -10, 0), label, value, _SECTION,
                        data_bold=data_bold)

    def _cell(self, p, rect, num, s, data_bold, *, show_delta=False) -> None:
        status = s.get("status")
        bg_key = {"best": "sec_best", "running": "sec_running",
                  "done": "sec_done"}.get(status, "sec_idle")
        inner = rect.adjusted(1, 1, -1, -1)
        rad = cell_radius(rect.height())
        if status in (None, "idle"):
            draw_dark_cell(p, inner, _SECTION, radius=rad)
        else:
            p.setPen(QPen(col("cell_border", _SECTION), 1))
            p.setBrush(col(bg_key, _SECTION))
            p.drawRoundedRect(inner, rad, rad)
        if s.get("active"):
            p.setPen(QPen(col("sec_running_edge", _SECTION), 1.6))
            p.setBrush(Qt.BrushStyle.NoBrush)
            p.drawRoundedRect(inner, rad, rad)
        p.setFont(tfont(rect.height() * 0.26, bold=False))
        p.setPen(col("sec_text", _SECTION))
        p.drawText(QRectF(rect.left(), rect.top() + rect.height() * 0.08,
                          rect.width(), rect.height() * 0.40),
                   Qt.AlignmentFlag.AlignCenter, f"S{num}")
        p.setFont(tabfont(rect.height() * 0.34, bold=data_bold))
        p.drawText(QRectF(rect.left(), rect.center().y() - rect.height() * 0.05,
                          rect.width(), rect.height() * 0.40),
                   Qt.AlignmentFlag.AlignCenter, sec(s.get("time")))
        if show_delta:
            delta = s.get("delta")
            if isinstance(delta, (int, float)) and abs(delta) >= 0.005:
                p.setFont(tfont(rect.height() * 0.22, bold=False))
                dc = col("slower", _SECTION) if delta > 0 else col("faster", _SECTION)
                p.setPen(dc)
                p.drawText(QRectF(rect.left(), rect.bottom() - rect.height() * 0.32,
                                  rect.width(), rect.height() * 0.28),
                           Qt.AlignmentFlag.AlignCenter, signed_delta(delta, 2))

File: overlay-rs/crates/gridglance-overlay/src/widgets/sector_timing.rs
//! Sector / lap timing widget — current, last, best + sector cells.

use super::WidgetCtx;
use crate::chrome::{draw_card, draw_dark_cell, full_rect, label, panel_pad};
use crate::telemetry::SectorCell;
use egui::{Align2, CornerRadius, Pos2, Rect, Stroke, Ui};

const SECTION: &str = "sector_timing";

/// Format seconds as `M:SS.mmm` (Python `clock`).
fn fmt_clock(sec: Option<f64>) -> String {
    match sec {
        Some(s) if s > 0.0 => {
            let m = (s / 60.0).floor() as i32;
            let rem = s - m as f64 * 60.0;
            format!("{m}:{rem:06.3}")
        }
        _ => "--:--.---".into(),
    }
}

/// Short sector split display (Python `sec`).
fn fmt_sec(sec: Option<f64>) -> String {
    match sec {
        Some(s) if s > 0.0 => format!("{s:.1}"),
        _ => "--.-".into(),
    }
}

fn signed_delta(d: f64, places: usize) -> String {
    if places == 2 {
        format!("{d:+.2}")
    } else {
        let sign = if d < 0.0 { '-' } else { '+' };
        format!("{sign}{:.prec$}", d.abs(), prec = places)
    }
}

fn cell_radius(h: f32) -> f32 {
    (h * 0.18).clamp(4.0, 10.0)
}

fn draw_edge_band(ui: &mut Ui, cfg: &crate::config::OverlayConfig, rect: Rect) {
    ui.painter().rect_filled(
        rect,
        CornerRadius::ZERO,
        cfg.color(SECTION, "header_bg", "#0b0e12bb"),
    );
    ui.painter().line_segment(
        [rect.left_bottom(), rect.right_bottom()],
        Stroke::new(1.0_f32, cfg.color(SECTION, "border", "#ffffff28")),
    );
}

fn draw_metric_pair(ui: &mut Ui, ctx: &WidgetCtx<'_>, rect: Rect, lab: &str, value: &str) {
    let scale = ctx.cfg.text_scale(SECTION);
    label(
        ui,
        Pos2::new(rect.left() + 10.0, rect.center().y),
        Align2::LEFT_CENTER,
        lab,
        (rect.height() * 0.38 * scale).clamp(9.0, 14.0),
        ctx.cfg.color(SECTION, "muted", "#8b93a1"),
        true,
    );
    label(
        ui,
        Pos2::new(rect.right() - 10.0, rect.center().y),
        Align2::RIGHT_CENTER,
        value,
        (rect.height() * 0.46 * scale).clamp(11.0, 18.0),
        ctx.cfg.color(SECTION, "text", "#f4f6f8"),
        true,
    );
}

fn paint_sector_cell(
    ui: &mut Ui,
    ctx: &WidgetCtx<'_>,
    rect: Rect,
    num: usize,
    cell: &SectorCell,
    show_delta: bool,
) {
    let scale = ctx.cfg.text_scale(SECTION);
    let inner = rect.shrink(1.0);
    let rad = cell_radius(rect.height());
    let status = cell.status.as_str();
    let bg_key = match status {
        "best" => ("sec_best", "#6b39c8"),
        "running" => ("sec_running", "#1d3a2a"),
        "done" => ("sec_done", "#22303f"),
        _ => ("sec_idle", "#161a20"),
    };
    if status == "idle" || status.is_empty() {
        draw_dark_cell(ui, ctx.cfg, SECTION, inner, rad);
    } else {
        ui.painter().rect_filled(
            inner,
            CornerRadius::same(rad as u8),
            ctx.cfg.color(SECTION, bg_key.0, bg_key.1),
        );
        ui.painter().rect_stroke(
            inner,
            CornerRadius::same(rad as u8),
            Stroke::new(1.0_f32, ctx.cfg.color(SECTION, "cell_border", "#ffffff28")),
            egui::StrokeKind::Inside,
        );
    }
    if cell.active {
        ui.painter().rect_stroke(
            inner,
            CornerRadius::same(rad as u8),
            Stroke::new(
                1.6_f32,
                ctx.cfg.color(SECTION, "sec_running_edge", "#46df7a"),
            ),
            egui::StrokeKind::Inside,
        );
    }
    let sec_text = ctx.cfg.color(SECTION, "sec_text", "#dfe3ea");
    label(
        ui,
        Pos2::new(rect.center().x, rect.top() + rect.height() * 0.28),
        Align2::CENTER_CENTER,
        &format!("S{num}"),
        (rect.height() * 0.26 * scale).clamp(9.0, 16.0),
        sec_text,
        false,
    );
    label(
        ui,
        Pos2::new(rect.center().x, rect.center().y + rect.height() * 0.12),
        Align2::CENTER_CENTER,
        &fmt_sec(cell.time),
        (rect.height() * 0.34 * scale).clamp(11.0, 20.0),
        sec_text,
        true,
    );
    if show_delta {
        if let Some(d) = cell.delta.filter(|d| d.abs() >= 0.005) {
            let dc = if d > 0.0 {
                ctx.cfg.color(SECTION, "slower", "#e23b3b")
            } else {
                ctx.cfg.color(SECTION, "faster", "#46df7a")
            };
            label(
                ui,
                Pos2::new(rect.center().x, rect.bottom() - rect.height() * 0.16),
                Align2::CENTER_CENTER,
                &signed_delta(d, 2),
                (rect.height() * 0.22 * scale).clamp(8.0, 14.0),
                dc,
                false,
            );
        }
    }
}

pub fn paint(ui: &mut Ui, ctx: &mut WidgetCtx<'_>) {
    let rect = full_rect(ui);
    let (card, _radius) = draw_card(ui, ctx.cfg, SECTION, rect);
    let pad = panel_pad(card.height());
    let h = card.height();
    let scale = ctx.cfg.text_scale(SECTION);
    let snap = &ctx.frame.sectors_ui;
    let show_pred = ctx.cfg.bool_key(SECTION, "show_predicted_lap", false);
    let show_delta = ctx.cfg.bool_key(SECTION, "show_sector_delta", false);

    let iw = card.width() - 2.0 * pad;
    let cur_h = if show_pred { h * 0.26 } else { h * 0.30 };
    label(
        ui,
        Pos2::new(card.center().x, card.top() + pad + cur_h * 0.5),
        Align2::CENTER_CENTER,
        &fmt_clock(snap.cur_lap),
        (cur_h * 0.72 * scale).clamp(18.0, 42.0),
        ctx.cfg.color(SECTION, "text", "#f4f6f8"),
        true,
    );

    if show_pred {
        if let Some(pred) = snap.predicted_lap.filter(|t| *t > 0.0) {
            label(
                ui,
                Pos2::new(card.center().x, card.top() + pad + cur_h * 0.95),
                Align2::CENTER_CENTER,
                &format!("Pred {}", fmt_clock(Some(pred))),
                (h * 0.09 * scale).clamp(10.0, 16.0),
                ctx.cfg.color(SECTION, "muted", "#8b93a1"),
                false,
            );
        }
    }

    let sub_top = card.top() + pad + if show_pred { h * 0.34 } else { h * 0.30 };
    let fixed_rh = ctx.cfg.f64_key(SECTION, "row_height_px", 0.0) as f32;
    let mut sub_h = if fixed_rh > 0.0 { fixed_rh } else { h * 0.18 };
    let max_frac = ctx.cfg.f64_key(SECTION, "max_row_height_frac", 0.0) as f32;
    if max_frac > 0.0 {
        sub_h = sub_h.min(h * max_frac);
    }
    sub_h = sub_h.max(18.0);
    let sub = Rect::from_min_size(Pos2::new(card.left() + pad, sub_top), egui::vec2(iw, sub_h));
    draw_edge_band(ui, ctx.cfg, sub);
    let half = sub.width() * 0.5;
    draw_metric_pair(
        ui,
        ctx,
        Rect::from_min_size(sub.left_top(), egui::vec2(half, sub.height())),
        "LAST",
        &fmt_clock(snap.last_lap),
    );
    draw_metric_pair(
        ui,
        ctx,
        Rect::from_min_size(
            Pos2::new(sub.left() + half, sub.top()),
            egui::vec2(half, sub.height()),
        ),
        "BEST",
        &fmt_clock(snap.best_lap),
    );

    let sectors = &snap.sectors;
    if sectors.is_empty() {
        return;
    }
    let top = sub.bottom() + h * 0.04;
    let ch = (card.bottom() - pad - top).max(24.0);
    let gap = iw * 0.03;
    let n = sectors.len() as f32;
    let cw = (iw - gap * (n - 1.0).max(0.0)) / n.max(1.0);
    let mut x = card.left() + pad;
    for (i, cell) in sectors.iter().enumerate() {
        let cell_rect = Rect::from_min_size(Pos2::new(x, top), egui::vec2(cw, ch));
        paint_sector_cell(ui, ctx, cell_rect, i + 1, cell, show_delta);
        x += cw + gap;
    }
}
